### PR TITLE
fix(cli): pass app options from command-line

### PR DESCRIPTION
**Pull request by aaron-kasten**
This PR fixes issue #2869 .  

## Scenario
Repository `aaron-kopia-repo-v1` is in process of an upgrade by another user, this user has acquired an upgrade lock on the repository.

### before application of this PR:
```
GOROOT=/opt/local/lib/go #gosetup
GOPATH=/Users/aalpar/devel/kopia/go #gosetup
repo connect s3 --bucket=aaron-kopia-repo-v1 --access-key=AK...FT --secret-access-key=Ne1O...78e --upgrade-no-block --password ******
... <will wait forever>....
^C
```

### after application of this PR:
```
GOROOT=/opt/local/lib/go #gosetup
GOPATH=/Users/aalpar/devel/kopia/go #gosetup
repo connect s3 --bucket=aaron-kopia-repo-v1 --access-key=AK...FT --secret-access-key=Ne1O...78e --upgrade-no-block --password ******
ERROR failed to open repository: repository upgrade in progress
ERROR error connecting to repository: repository upgrade in progress

Process finished with the exit code 1


```

### CODE REVIEW

File: repo/connect.go [hunk 1]
```diff
@@ -25,9 +25,9 @@ type ConnectOptions struct {
 var ErrRepositoryNotInitialized = errors.Errorf("repository not initialized in the provided storage")
 
 // Connect connects to the repository in the specified storage and persists the configuration and credentials in the file provided.
-func Connect(ctx context.Context, configFile string, st blob.Storage, password string, opt *ConnectOptions) error {
-	if opt == nil {
-		opt = &ConnectOptions{}
+func Connect(ctx context.Context, configFile string, st blob.Storage, password string, options *Options, connectOptions *ConnectOptions) error {
```
Comment (jkowalski) [inline repo/connect.go:28]: Can we simply add Options as a field inside ConnectOptions?

Comment (aaron-kasten) [reply repo/connect.go:28]: @jkowalski see if the recent commit is a fix.  Combining the two structures closer to the command line causes problems with the use case of each (one is for configuration only, the other is for general command-line flags) and visibility.

File: repo/blob/filesystem/filesystem_storage.go [hunk 2]
```diff
@@ -306,7 +308,7 @@ func (fs *fsStorage) TouchBlob(ctx context.Context, blobID blob.ID, threshold ti
 
 		//nolint:wrapcheck
 		return osi.Chtimes(path, n, n)
-	}, fs.Impl.(*fsImpl).isRetriable) //nolint:forcetypeassert
+	}, fsimp.isRetriable)
```
Comment (adowair) [inline repo/blob/filesystem/filesystem_storage.go:311]: Forgive my ignorance, what is the relevance of this change here?

File: tests/end_to_end_test/repository_connect_test.go [hunk 3]
```diff
@@ -77,6 +79,19 @@ func TestReconnect(t *testing.T) {
...
+	t.Parallel()
+
+	runner := testenv.NewInProcRunner(t)
+	e := testenv.NewCLITest(t, testenv.RepoFormatNotImportant, runner)
+
+	e.RunAndExpectSuccess(t, "repo", "create", "filesystem", "--path", e.RepoDir)
+	e.RunAndExpectSuccess(t, "repo", "disconnect")
+	filesystem.New(context.Background(), &filesystem.Options{Path: e.RepoDir}, false)
+	e.RunAndExpectSuccess(t, "repo", "connect", "filesystem", "--path", e.RepoDir)
+	e.RunAndExpectSuccess(t, "repo", "status")
```
Comment (adowair) [inline tests/end_to_end_test/repository_connect_test.go:92]: I think we need to cover the targeted scenario as well, no?
```suggestion
  e := testenv.NewCLITest(t, testenv.RepoFormatNotImportant, runner)
  e.Environment["KOPIA_UPGRADE_LOCK_ENABLED"] = "1"

	e.RunAndExpectSuccess(t, "repo", "create", "filesystem", "--path", e.RepoDir, "--format-version", "2")
	e.RunAndExpectSuccess(t, "repo", "upgrade", "begin", "--lock-only", "--upgrade-owner-id", "admin")
	e.RunAndExpectSuccess(t, "repo", "disconnect")
	e.RunAndExpectFailure(t, "repo", "connect", "filesystem", "--path", e.RepoDir, "--upgrade-no-block")
```

**Review by adowair — CHANGES_REQUESTED** [review]
Thanks @aaron-kasten . I think we need to add more test coverage. In addition, a small question about some code I don't understand.

File: repo/blob/filesystem/filesystem_storage.go (hunk 2 above)
Comment (aaron-kasten) [reply repo/blob/filesystem/filesystem_storage.go:311]: the assertion doesn't do anything - so I removed it.

File: repo/connect.go [hunk 4]
```diff
@@ -18,16 +18,18 @@ type ConnectOptions struct {
 	ClientOptions
 
 	content.CachingOptions
+
+	Options *Options `json:"-"`
```
[the remark this answers is not in the document]
Comment (aaron-kasten) [reply repo/connect.go:22]: @adowair JSON serializer will create a default entry if not explicitly left out.  Change was made at the request of @jkowalski

[the remark this answers is not in the document]
Comment (aaron-kasten) [reply repo/connect.go:22]: https://pkg.go.dev/encoding/json

File: repo/connect.go (hunk 1 above)
Comment (Shrekster) [reply repo/connect.go]: Had a quick chat with Aaron, It is desirable to have a single options parser for CLI. Does embedding the structure into a section options struct for connect make them separate ?

**Pull request closed without merging**